proposals/0723-qualified-strings.rst

```diff
@@ -0,0 +1,282 @@
+
+See also:
+
+* `QualifiedLists <https://github.com/ghc-proposals/ghc-proposals/pull/724>`_
+* `QualifiedNumerics <https://github.com/ghc-proposals/ghc-proposals/pull/725>`_
+
+Motivation
+----------
+
+Problems with Type Class-driven overloading
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``OverloadedStrings`` works by desugaring to a type-class-overloaded function. That leads to certain general shortcomings:
+
+* It is a module-wide setting
+
+  * Anecdotally, people would rather avoid ``OverloadedStrings`` than deal with overloaded lists/strings in the entire module.
+
+  * It's possible that this is one reason these extensions aren't defaults in GHC202X language editions, despite being in GHC for a long time.
+
+* Type inference ambiguity.
+
+  * Consider the following code:
+
+    ::
+
+      output :: IsString s => s -> IO ()
+
+      main = do
+        output "hello"
+        output "world"
+
+        -- output $ T.replace " " "_" input
+
+    This originally works with no extensions, due to the string literals being typed to concrete ``String``. But say the developer wants to add a call to ``T.replace`` and use ``Text`` literals; adding ``OverloadedStrings`` would cause ambiguity to the existing locations because they are now no longer concretely ``String``.
+
+This proposal would allow using a module qualifier to say precisely which function to desugar to, rather than using type classes, in a similar manner as ``-XQualifiedDo``. This would allow writing the previous code as
+
+::
+
+  {-# LANGUAGE QualifiedStrings #-}
+
+  main = do
+    output "hello"
+    output "world"
+
+    output $ T.replace T." " T."_" input
+
+The existing locations would continue working as ``String``, while the new line would unambiguously desugar to ``T.replace (T.pack " ") (T.pack "_") input``.
+
+Proposed Change Specification
+-----------------------------
+
+Introduce ``-XQualifiedStrings`` that desugars literal string syntax to function calls in a similar way to ``-XQualifiedDo`` (`docs <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/qualified_do.html>`_, `proposal <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0216-qualified-do.rst>`_).
+
+As long as the desugared expressions/patterns type check, users are free to define these functions however they want. No whitespace is allowed between the ``.`` and the module name / literal.
+
+Currently, string literals have the following desugaring:
+
+.. list-table::
+    :align: left
+
+    * - **Expression**
+      - **Enabled extensions**
+      - **Desugared expression syntax**
+    * - ``"hello"``
+      -
+      - ``"hello"``
+    * - ``"hello"``
+      - ``-XOverloadedStrings``
+      - ``GHC.Exts.fromString "hello"``
```

Comment:
i think it would be nice if this table also had a -XMultilineStrings example, just to make it clear that the `"""..."""` syntax isnt added by this proposal